tp/tp.py

```diff
@@ -136,6 +138,21 @@ def refresh(self):
                              self.previous_as_count, self.managed_by_autoscale())
             self.previous_as_count = self.managed_by_autoscale()
 
+    def is_ec2_state_running(self, instance_id):
+        def ec2_state(i_id):
```

Comment:
You use this method only once. I don't see a good enough motivation to define this inside method `ec2_state`. Moreover, it's a bit weird defining a method inside another method and reference a global variable (`self.ec2`). I would simply use `self.ec2.get_all_instance_status(instance_ids=[instance_id])` instead of `ec2_state(instance_id)`.